io/zenoh-links/zenoh-link-udp/src/unicast.rs

```diff
@@ -347,50 +337,132 @@ impl LinkManagerUnicastTrait for LinkManagerUnicastUdp {
             e
         })?;
 
-        // Update the endpoint locator address
-        assert!(endpoint.set_addr(&format!("{}", local_addr)));
+        Ok((socket, local_addr))
+    }
+}
 
-        // Spawn the accept loop for the listener
-        let active = Arc::new(AtomicBool::new(true));
-        let signal = Signal::new();
+#[async_trait]
+impl LinkManagerUnicastTrait for LinkManagerUnicastUdp {
+    async fn new_link(&self, endpoint: EndPoint) -> ZResult<LinkUnicast> {
+        let dst_addrs = get_udp_addrs(&endpoint.locator)
+            .await?
+            .drain(..)
+            .filter(|a| !a.ip().is_multicast())
+            .collect::<Vec<SocketAddr>>();
+
+        let mut errs: Vec<ZError> = vec![];
+        for da in dst_addrs.iter() {
+            match self.new_link_inner(da).await {
+                Ok((socket, src_addr, dst_addr)) => {
+                    // Create UDP link
+                    let link = Arc::new(LinkUnicastUdp::new(
+                        src_addr,
+                        dst_addr,
+                        LinkUnicastUdpVariant::Connected(LinkUnicastUdpConnected {
+                            socket: Arc::new(socket),
+                        }),
+                    ));
 
-        let c_active = active.clone();
-        let c_signal = signal.clone();
-        let c_manager = self.manager.clone();
-        let c_listeners = self.listeners.clone();
-        let c_addr = local_addr;
-        let handle = task::spawn(async move {
-            // Wait for the accept loop to terminate
-            let res = accept_read_task(socket, c_active, c_signal, c_manager).await;
-            zwrite!(c_listeners).remove(&c_addr);
-            res
-        });
+                    return Ok(LinkUnicast(link));
+                }
+                Err(e) => {
+                    errs.push(e);
+                }
+            }
+        }
 
-        let locator = endpoint.locator.clone();
-        let listener = ListenerUnicastUdp::new(endpoint, active, signal, handle);
-        // Update the list of active listeners on the manager
-        zwrite!(self.listeners).insert(local_addr, listener);
+        if errs.is_empty() {
+            errs.push(zerror!("No unicast addresses available").into());
+        }
 
-        Ok(locator)
+        bail!(
+            "Can not create a new UDP link bound to {}: {:?}",
+            endpoint,
+            errs
+        )
+    }
+
+    async fn new_listener(&self, mut endpoint: EndPoint) -> ZResult<Locator> {
+        let addrs = get_udp_addrs(&endpoint.locator).await?;
```

Comment:
Shouldn't multicast addresses be filtered out here as well ?

Reply:
Fixed.